### PR TITLE
Fix BSplineCurve2d.line intersections()

### LINKED ISSUE #535: strange behavior cut_by_line
* **I'm submitting a ...**
  - [x] bug report
  - [ ] feature request

* **What is the current behavior?**

Contour2D.cut_by_line(line) detects 3 intersections points, but it should be 2. See image bellow:
[cut_by_line_problem.zip](https://github.com/Dessia-tech/volmdlr/files/10013569/cut_by_line_problem.zip)
[intersection_problem.zip](https://github.com/Dessia-tech/volmdlr/files/10040319/intersection_problem.zip)

image
image
* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** Avoid reference to other packages

* **What is the expected behavior?**
Expected to have only two intersections points

* **What is the motivation / use case for changing the behavior?**

* **Possible fixes**

* **Please tell us about your environment:**
  
  - branch: 2_step_to_volmdlr
  - commit:  
  - python version: 3.10

### PR DESCRIPTION
**Pull request by WirajanDASILVA**
* **Please check if the PR fulfills these requirements**
- [X] The commit messages follows our guidelines (explicit, in english)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [X] Bug fix:
- [ ] New features:
- [ ] Documentation
- [ ] Optimization

* **Other information**:
closes #535

### CODE REVIEW

File: volmdlr/edges.py [hunk 1]
```diff
@@ -819,6 +819,14 @@ def sort_points_along_line(self, points: List[volmdlr.Point2D]) -> List[
...
         return list_points
 
+    def point_belongs(self, point2d, abs_tol: float = 1e-6):
+        """
+        Verifies if the point2d belongs to the line
+        :param point2d: point to be verified
+        :return: True if point belongs to line, False otherwise
+        """
+        return math.isclose(self.point_distance(point2d), 0, abs_tol=abs_tol)
+
```
Comment (WirajanDASILVA) [inline volmdlr/edges.py:829]: Missing Unittests for this new method

File: volmdlr/edges.py [hunk 2]
```diff
@@ -1202,15 +1212,10 @@ def line_intersections(self, line: Line2D):
...
-            vector3 = self.end - line.point1
-            vector4 = self.end - line.point2
-            if math.isclose(vector1.cross(vector2), 0, abs_tol=1e-6):
-                return [self.start]
-            if math.isclose(vector3.cross(vector4), 0, abs_tol=1e-6):
-                return [self.end]
+        if line.point_belongs(self.start):
+            return [self.start]
+        if line.point_belongs(self.end):
+            return [self.end]
```
Comment (WirajanDASILVA) [inline volmdlr/edges.py:1218]: add unittest for LineSegment2D.line_intersections for all different scenarios

File: tests/edges/test_line2d.py [hunk 3]
```diff
@@ -11,6 +11,13 @@ def test_point_distance(self):
...
+    def test_point_belongs(self):
+        line = edges.Line2D(volmdlr.O2D, volmdlr.Point2D(1, 1))
+        point1 = volmdlr.Point2D(2, 2)
+        point2 = volmdlr.Point2D(1, 2)
+        self.assertTrue(line.point_belongs(point1))
+        self.assertFalse(line.point_belongs(point2))
+
 
 if __name__ == '__main__':
     unittest.main()
```
Comment (WirajanDASILVA) [inline tests/edges/test_line2d.py:23]:
```suggestion
    unittest.main()

```

File: tests/edges/test_bsplinecurve.py [hunk 4]
```diff
@@ -14,6 +16,15 @@ def test_abscissa(self):
...
+        line = edges.Line2D(volmdlr.Point2D(1.263163105753452, -0.002645572020392778),
+                            volmdlr.Point2D(1.263163105753452, -0.001820963841291406))
+
+        line_intersections = bspline_curve2d.line_intersections(line)
+        self.assertEqual(len(line_intersections), 1)
+        self.assertEqual(line_intersections[0], volmdlr.Point2D(1.2631631057526727, -0.0026450894385881708))
+
 
 if __name__ == '__main__':
     unittest.main()
```
Comment (WirajanDASILVA) [inline tests/edges/test_bsplinecurve.py:30]:
```suggestion
    unittest.main()

```

File: volmdlr/edges.py [hunk 5]
```diff
@@ -956,6 +956,40 @@ def create_tangent_circle(self, point, other_line):
...
+        points.remove(most_distant_point)
+        distances_to_reference_point = {}
+        for point in points:
+            distances_to_reference_point[point] = \
+                most_distant_point.point_distance(point)
+        distances_to_reference_point = dict(
+            sorted(distances_to_reference_point.items(),
+                   key=lambda item: item[1]))
+        list_points.extend(list(distances_to_reference_point.keys()))
+        return list_points
```
Comment (WirajanDASILVA) [inline volmdlr/edges.py:983]:
```suggestion
```

**Review by masfaraud — APPROVED** [review]

**Pull request closed without merging**